flink-core/src/main/java/org/apache/flink/configuration/ConfigOptionsDocGenerator.java

```diff
@@ -0,0 +1,279 @@
+			tables.add(Tuple2.<ConfigGroup, String>of(null, toHTMLTable(allOptions)));
+		}
+		return tables;
+	}
+
+	/**
+	 * Method used to generated documentation entries containing tables of configuration options with default value and
+	 * description. Each found class matching a pattern *Options.java will results in a separate file with a configuration
+	 * table.
+	 *
+	 * @param args first argument is output path for the generated files, second argument is full package name containing
+	 *             classes with {@link ConfigOption}
+	 */
+	public static void main(String[] args) throws IOException, ClassNotFoundException {
+		final String outputPath = args[0];
+		final String packageName = args[1];
+
+		final Path configDir = Paths.get("../src/main/java", packageName.replaceAll("\\.", "/"));
+
+		final Pattern p = Pattern.compile("(([a-zA-Z]*)(Options))\\.java");
+		try (DirectoryStream<Path> stream = Files.newDirectoryStream(configDir, "*Options.java")) {
+			for (Path entry: stream) {
+				final String fileName = entry.getFileName().toString();
+				final Matcher matcher = p.matcher(fileName);
+				if (!fileName.equals("ConfigOptions.java") && matcher.matches()) {
+					final Class<?> optionsClass = Class.forName(packageName + "." + matcher.group(1));
+					List<Tuple2<ConfigGroup, String>> tables = generateTablesForClass(optionsClass);
+					for (Tuple2<ConfigGroup, String> group : tables) {
+
+						String name = group.f0 == null
+							? matcher.group(2).replaceAll("(.)(\\p{Upper})", "$1_$2").toLowerCase()
+							: group.f0.name().replaceAll("(.)(\\p{Upper})", "$1_$2").toLowerCase();
+
+						final String outputFile = name + "_configuration.html";
+						Files.write(Paths.get(outputPath, outputFile), group.f1.getBytes(StandardCharsets.UTF_8));
+					}
+				}
+			}
+		}
+	}
+
+	private static class Tree {
+		private final Node root = new Node();
+
+		private final Node defaultGroup = new Node();
+
+		public Tree() {
+		}
+
+		private static class Node {
+			private final List<ConfigOption> configOptions = new ArrayList<>();
+			private final Map<String, Node> children = new HashMap<>();
+			private boolean terminal = false;
+
+			Node addChild(String keyComponent) {
+				Node child = children.get(keyComponent);
+				if (child == null) {
+					child = new Node();
+					children.put(keyComponent, child);
+				}
+				return child;
+			}
+
+			Node findChild(String keyComponent) {
+				Node child = children.get(keyComponent);
+				if (child == null) {
+					return this;
+				}
+				return child;
+			}
+
+			void addOption(ConfigOption option) {
+				configOptions.add(option);
+			}
+
+			public boolean isTerminal() {
+				return terminal;
+			}
+
+			public void setTerminal() {
+				this.terminal = true;
+			}
+
+			public List<ConfigOption> getConfigOptions() {
+				return configOptions;
+			}
+		}
+
+		public void add(ConfigOption<?> option) {
+			findNode(option.key())
+				.addOption(option);
+		}
+
+		public void set(ConfigGroup group) {
```

Comment:
We could move this into the constructor by adding a `List<ConfigGroup>` argument. This would make it more obvious that they are what defines the structure of the tree.